### PR TITLE
fix #1152 crash by putting mutex on sendRetryRequest()

### LINKED ISSUE #1152: [MD] Baileys crashes on 500+ missed messages
I spammed myself over the night with messages because I recently saw a problem when an instance is offline for a few hours.

I sent a bit more messages than I intended to do but in my case it went the following:

`12037 offline messages/notifications`

it results in a lot of unknown message keys and therefore retry requests. After a while the following pops up:

```
src/Socket/socket.ts:328:17

Code":408,"error":"Request Time-out","message":"Timed Out"},"headers":{}}},"msg":"connection closed"}
connection update {
  connection: 'close',
  lastDisconnect: {
    error: Error: Timed Out
        at /src/Utils/generics.ts:167:13
        at runNextTicks (node:internal/process/task_queues:61:5)
        at listOnTimeout (node:internal/timers:526:9)
        at processTimers (node:internal/timers:500:7) 
    }
```

the connection automatically restarts. then, after a few more seconds, it does:

```
[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "1006".] {
  code: 'ERR_UNHANDLED_REJECTION'
}
```

but without a trace to where it's happening. if I search for `1006` I find a line in the `ws` package:

`this._closeCode = 1006;`

I don't know how the website does it but maybe we have to either
-  batch request missing messages for retries, 
- **or make these retries throttle a bit to only doing 50 retries at once?** we could also increase the timeout but this does not sound like the right way to do it. throttling them may also reduce memory footprint which can be an issue as pointed out by #1119

I noticed the stack for the former is `src/Socket/socket.ts:328:17` (startKeepAliveRequest \ waitForMessage \ promiseTimeout) so maybe 
- the input load on the process is so high that the server response to the ping either never arrives to be put on the stack
- or it actually reaches the stack but Baileys cannot work through it in time due to all the other messages from the retry requests coming in

in that case we could either prioritize the keepAlive or what I would prefer to set a limit for the amount of pending requests before the old ones have been responded to so it won't run under full load potentially leading to other issues down the line.

the way it is right now is probably the fastest way since it's using a max of CPU and RAM in a short amount of time however putting a layer of queuing there would probably be much safer with only a percent of decline in performance since we can still have dozens of requests taking place at the same time, it's just that when it hits "hundreds" the issues seem to arise.

do you have any thoughts on this @adiwajshing ?

### PR DESCRIPTION
**Pull request by HUGEIT**
this supposedly resolves a crash I described in #1152 that happens when a lot of "Waiting for this message" messages have to be retried in Baileys. in my live tests, one request was dealt with after another.

I still think this could be parallelized a little but this is a step forward imo

```
48649cd8 , start of sending retry request
48649cd8 , before assertingPreKeys
48649cd8 , before sendNode
48649cd8 , after sendNode <-- this should be the last call
9f04a3c8 , start of sending retry request
9f04a3c8 , before assertingPreKeys
9f04a3c8 , before sendNode
9f04a3c8 , after sendNode <-- this should be the last call
b79bc306 , start of sending retry request
b79bc306 , before assertingPreKeys
b79bc306 , before sendNode
b79bc306 , after sendNode <-- this should be the last call
bc7a8391 , start of sending retry request
bc7a8391 , before assertingPreKeys
bc7a8391 , before sendNode
bc7a8391 , after sendNode <-- this should be the last call
(no crossovers in there although they all fired around the same time)
```

thank you adi for suggesting that

### CODE REVIEW

**Comment by HUGEIT** [comment]
testing went great :) ✔️

I have just restarted a session with 1120 retries and Baileys did not crash - since this implementation does not wait for a response it performs quite well in terms of time. I am still concerned about blocking the receiving end of the socket (because I thought that was the issue of the timeout not arriving) but if what the mutex does slows down the key generation, sending etc. already suffices this sounds like an ideal solution.

File: src/Socket/messages-recv.ts [hunk 1]
```diff
@@ -6,13 +6,16 @@ import { decodeMessageStanza, downloadAndProcessHistorySyncNotification, encodeB
...
 import { extractGroupMetadata } from './groups'
+import makeMutex from '../Utils/make-mutex'
 
 const STATUS_MAP: { [_: string]: proto.WebMessageInfo.WebMessageInfoStatus } = {
 	'played': proto.WebMessageInfo.WebMessageInfoStatus.PLAYED,
 	'read': proto.WebMessageInfo.WebMessageInfoStatus.READ,
 	'read-self': proto.WebMessageInfo.WebMessageInfoStatus.READ
 }
 
+const retryMutex = makeMutex()
```
Comment (adiwajshing) [inline src/Socket/messages-recv.ts:17]: I feel this should be inside the connection object

**Comment by adiwajshing** [comment]
Looks mostly good @HUGEIT 

I recently mutexed all operations on individual chats to maintain the correct order of things (https://github.com/adiwajshing/Baileys/commit/1f2a6641f32d563ba8707997e6a13e60b94237fe). Can you try the master branch without this PR and see if that works? Because I've a feeling this change might be redundant with the other change I did.

**Comment by HUGEIT** [comment]
hey @adiwajshing thank you for your comment and your commit. but I think that can still cause an overload if retries happen in multiple chats at once.

we still need the the "global" mutex I think, with your latest commit without a specific retry-mutex we face the following behavior in a live environment, which may be an improvement but already looks very dangerous again (see the log at the bottom).

can you tell me what you mean by the `I feel this should be inside the connection object`? Do you mean the `sock`?

Log with your commit:

```
**  e41 , start of sending retry request
**  e41 , before assertingPreKeys
**  861 , start of sending retry request
**  861 , before assertingPreKeys
**  522 , start of sending retry request
**  522 , before assertingPreKeys
**  462 , start of sending retry request
**  462 , before assertingPreKeys
**  92f , start of sending retry request
**  92f , before assertingPreKeys
**  387 , start of sending retry request
**  387 , before assertingPreKeys
**  fae , start of sending retry request
**  fae , before assertingPreKeys
**  88a , start of sending retry request
**  88a , before assertingPreKeys
**  310 , start of sending retry request
**  310 , before assertingPreKeys
**  5cf , start of sending retry request
**  5cf , before assertingPreKeys
**  1ea , start of sending retry request
**  1ea , before assertingPreKeys
**  89f , start of sending retry request
**  89f , before assertingPreKeys
**  312 , start of sending retry request
**  312 , before assertingPreKeys
**  d53 , start of sending retry request
**  d53 , before assertingPreKeys
**  830 , start of sending retry request
**  830 , before assertingPreKeys
**  237 , start of sending retry request
**  237 , before assertingPreKeys
**  d93 , start of sending retry request
**  d93 , before assertingPreKeys
**  20e , start of sending retry request
**  20e , before assertingPreKeys
**  5d0 , start of sending retry request
**  5d0 , before assertingPreKeys
**  34a , start of sending retry request
**  34a , before assertingPreKeys
**  6ec , start of sending retry request
**  6ec , before assertingPreKeys
**  392 , start of sending retry request
**  392 , before assertingPreKeys
**  786 , start of sending retry request
**  786 , before assertingPreKeys
**  a67 , start of sending retry request
**  a67 , before assertingPreKeys
**  e41 , before sendNode
**  861 , before sendNode
**  522 , before sendNode
**  462 , before sendNode
**  92f , before sendNode
**  387 , before sendNode
**  fae , before sendNode
**  88a , before sendNode
**  310 , before sendNode
**  5cf , before sendNode
**  1ea , before sendNode
**  89f , before sendNode
**  312 , before sendNode
**  d53 , before sendNode
**  830 , before sendNode
**  237 , before sendNode
**  d93 , before sendNode
**  20e , before sendNode
**  5d0 , before sendNode
**  34a , before sendNode
**  6ec , before sendNode
**  392 , before sendNode
**  786 , before sendNode
**  a67 , before sendNode
**  1ec , start of sending retry request
**  1ec , before assertingPreKeys
**  06a , start of sending retry request
**  06a , before assertingPreKeys
**  7bc , start of sending retry request
**  7bc , before assertingPreKeys
**  ee8 , start of sending retry request
**  ee8 , before assertingPreKeys
**  561 , start of sending retry request
**  561 , before assertingPreKeys
**  e3d , start of sending retry request
**  e3d , before assertingPreKeys
**  5df , start of sending retry request
**  5df , before assertingPreKeys
**  213 , start of sending retry request
**  213 , before assertingPreKeys
**  d60 , start of sending retry request
**  d60 , before assertingPreKeys
**  e74 , start of sending retry request
**  e74 , before assertingPreKeys
**  60e , start of sending retry request
**  60e , before assertingPreKeys
**  007 , start of sending retry request
**  007 , before assertingPreKeys
**  9c1 , start of sending retry request
**  9c1 , before assertingPreKeys
**  988 , start of sending retry request
**  988 , before assertingPreKeys
**  a9e , start of sending retry request
**  a9e , before assertingPreKeys
**  06f , start of sending retry request
**  06f , before assertingPreKeys
**  de1 , start of sending retry request
**  de1 , before assertingPreKeys
**  9b8 , start of sending retry request
**  9b8 , before assertingPreKeys
**  583 , start of sending retry request
**  583 , before assertingPreKeys
**  71f , start of sending retry request
**  71f , before assertingPreKeys
**  480 , start of sending retry request
**  480 , before assertingPreKeys
**  54a , start of sending retry request
**  54a , before assertingPreKeys
**  e25 , start of sending retry request
**  e25 , before assertingPreKeys
**  c2c , start of sending retry request
**  c2c , before assertingPreKeys
**  322 , start of sending retry request
**  322 , before assertingPreKeys
**  8e4 , start of sending retry request
**  8e4 , before assertingPreKeys
**  9bc , start of sending retry request
**  9bc , before assertingPreKeys
**  303 , start of sending retry request
**  303 , before assertingPreKeys
**  11e , start of sending retry request
**  11e , before assertingPreKeys
**  735 , start of sending retry request
**  735 , before assertingPreKeys
**  ea6 , start of sending retry request
**  ea6 , before assertingPreKeys
**  6c5 , start of sending retry request
**  6c5 , before assertingPreKeys
**  003 , start of sending retry request
**  003 , before assertingPreKeys
**  1ec , before sendNode
**  06a , before sendNode
**  7bc , before sendNode
**  ee8 , before sendNode
**  561 , before sendNode
**  e3d , before sendNode
**  5df , before sendNode
**  213 , before sendNode
**  d60 , before sendNode
**  e74 , before sendNode
**  60e , before sendNode
**  007 , before sendNode
**  9c1 , before sendNode
**  988 , before sendNode
**  a9e , before sendNode
**  06f , before sendNode
**  de1 , before sendNode
**  9b8 , before sendNode
**  583 , before sendNode
**  71f , before sendNode
**  480 , before sendNode
**  54a , before sendNode
**  e25 , before sendNode
**  c2c , before sendNode
**  322 , before sendNode
**  8e4 , before sendNode
**  9bc , before sendNode
**  303 , before sendNode
**  11e , before sendNode
**  735 , before sendNode
**  ea6 , before sendNode
**  6c5 , before sendNode
**  003 , before sendNode
**  e41 , after sendNode <-- this should be the last call
**  861 , after sendNode <-- this should be the last call
**  522 , after sendNode <-- this should be the last call
**  462 , after sendNode <-- this should be the last call
**  92f , after sendNode <-- this should be the last call
**  387 , after sendNode <-- this should be the last call
**  fae , after sendNode <-- this should be the last call
**  88a , after sendNode <-- this should be the last call
**  310 , after sendNode <-- this should be the last call
**  5cf , after sendNode <-- this should be the last call
**  1ea , after sendNode <-- this should be the last call
**  89f , after sendNode <-- this should be the last call
**  312 , after sendNode <-- this should be the last call
**  d53 , after sendNode <-- this should be the last call
**  830 , after sendNode <-- this should be the last call
**  237 , after sendNode <-- this should be the last call
**  d93 , after sendNode <-- this should be the last call
**  20e , after sendNode <-- this should be the last call
**  5d0 , after sendNode <-- this should be the last call
**  34a , after sendNode <-- this should be the last call
**  6ec , after sendNode <-- this should be the last call
**  392 , after sendNode <-- this should be the last call
**  786 , after sendNode <-- this should be the last call
**  a67 , after sendNode <-- this should be the last call
**  1ec , after sendNode <-- this should be the last call
**  06a , after sendNode <-- this should be the last call
**  7bc , after sendNode <-- this should be the last call
**  ee8 , after sendNode <-- this should be the last call
**  561 , after sendNode <-- this should be the last call
**  e3d , after sendNode <-- this should be the last call
**  5df , after sendNode <-- this should be the last call
**  213 , after sendNode <-- this should be the last call
**  d60 , after sendNode <-- this should be the last call
**  e74 , after sendNode <-- this should be the last call
**  60e , after sendNode <-- this should be the last call
**  007 , after sendNode <-- this should be the last call
**  9c1 , after sendNode <-- this should be the last call
**  988 , after sendNode <-- this should be the last call
**  a9e , after sendNode <-- this should be the last call
**  06f , after sendNode <-- this should be the last call
**  de1 , after sendNode <-- this should be the last call
**  9b8 , after sendNode <-- this should be the last call
**  583 , after sendNode <-- this should be the last call
**  71f , after sendNode <-- this should be the last call
**  480 , after sendNode <-- this should be the last call
**  54a , after sendNode <-- this should be the last call
**  e25 , after sendNode <-- this should be the last call
**  c2c , after sendNode <-- this should be the last call
**  322 , after sendNode <-- this should be the last call
**  8e4 , after sendNode <-- this should be the last call
**  9bc , after sendNode <-- this should be the last call
**  303 , after sendNode <-- this should be the last call
**  11e , after sendNode <-- this should be the last call
**  735 , after sendNode <-- this should be the last call
**  ea6 , after sendNode <-- this should be the last call
**  6c5 , after sendNode <-- this should be the last call
**  003 , after sendNode <-- this should be the last call
```

**Comment by adiwajshing** [comment]
> Do you mean the sock?

Yes

**Comment by adiwajshing** [comment]
@HUGEIT I'm not entirely sure what's happening with the logs -- is it still crashing?

**Comment by HUGEIT** [comment]
@adiwajshing I cannot tell whether it is crashing until I have collected enough messages in a session. the only thing I can confirm so far is that with my PR there is no crash with 1000 retries. 

without "my" mutex at your current commit, if 1000 retries come from 1000 chats * 1 retries we might see a crash again if we do not have my "second layer mutex" on top of your recent commit.

sorry if I explain myself badly, it is hard to validate your current commit against a possible crash in a live environment because it is now impossible to reproduce with a single chat. but I can, in theory, see how a chat-exclusive mutex may not be enough to prevent the socket from over-loading with requests again if we deal with multiple chats.

the log is just to showcase that the function enters in multiple subroutines at once and potentially runs a lot of these requests in parallel, potentially leading to the crash again. for comparison please check the first comment where we have the mutex global across all chats so the function is only running once at a time

**Comment by adiwajshing** [comment]
@HUGEIT that's fair, just make the mutex part of the socket object -- and we're good to go with this

**Comment by HUGEIT** [comment]
ok thank you, I will make a new PR in a moment

**Comment by HUGEIT** [comment]
@adiwajshing can you take a moment and check if I messed up something there? https://github.com/HUGEIT/Baileys/commit/187484d6493ce1d4fd113fbc00ee881ba867c2e0

for some reason, with my mutex on top of your latest commits, I see the same crashing behavior as before, it will not wait for the sending of the other one to finish. can you spot a potential conflict of the mutexes in my commit? 

the first few work as expected (maybe by coincidence tho)

```
378 , start of sending retry request
378 , before assertingPreKeys
378 , before sendNode
378 , after sendNode <-- this should be the last call
641 , start of sending retry request
641 , before assertingPreKeys
641 , before sendNode
...
```

but then when there are masses of requests, we see the same non-waiting behavior as before:

```
048 , start of sending retry request
048 , before assertingPreKeys
048 , before sendNode
df2 , start of sending retry request
df2 , before assertingPreKeys
df2 , before sendNode
7e9 , start of sending retry request
7e9 , before assertingPreKeys
7e9 , before sendNode

(eventual 428 error)
```

**Comment by adiwajshing** [comment]
No, something is off -- this code should work. You can actually just keep the retryMutex inside the messages-recv.ts file as it does not need to be exposed to the world.

Also, can you open a PR please, much easier to review the code.

**Comment by HUGEIT** [comment]
sorry if I sound stupid but I do not understand the combination of 

`just make the mutex part of the socket object`

and 

`You can actually just keep the retryMutex inside the messages-recv.ts file` (as it was before)

where should I put it then? or do you mean that you changed your mind? if so, I will happily place it as it was before (when it was working at least haha)


apart from that, I just noticed that you already merged MD to master, I had my head still on MD.


I agree about the PR, but I did not want to submit a faulty PR to you.

**Comment by adiwajshing** [comment]
@HUGEIT no worries, so every file can have its own set of private variables (if you open `messages-recv.ts` -- it has a couple of objects used to manage the state that we do not need to expose to the rest of the world).

Also, your PR can just be a draft till the problem is resolved -- it can have bugs initially, no problem.

**Comment by HUGEIT** [comment]
I think now I understood what you meant - I declared it next to `STATUS_MAP` but instead you wanted me to put it in the block where things like `msgRetryMap` are located. I don't really understand TypeScript inheritance but I think my updated PR #1184 does what you mean now

**Pull request closed without merging**